### PR TITLE
Make MultiHeadAttention use masks from query and value tensors

### PR DESCRIPTION
**Pull request by mattsoulanille**
Add optional [Keras masks](https://www.tensorflow.org/guide/keras/understanding_masking_and_padding) to tfjs tensors. Enable them for tfjs-layers on layers that emit them. Use the masks of query and value input tensors in MultiHeadAttention to compute the correct mask automatically.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

### CODE REVIEW

File: tfjs-layers/src/engine/topology.ts [hunk 1]
```diff
@@ -1018,7 +1018,31 @@ export abstract class Layer extends serialization.Serializable {
...
+          const outputMask = this.computeMask(inputs as Tensor | Tensor[]);
+          if (output instanceof Array && outputMask instanceof Array) {
+            if (output.length !== outputMask.length) {
+              throw new Error(`${this.name} output ${output.length} tensors `
+                + `but ${outputMask.length} masks for those tensors`);
+            }
+            for (let i = 0; i < output.length; i++) {
+              output[i].kerasMask = outputMask[i];
+            }
+          } else if (outputMask instanceof Array) {
```
Comment (pyu10055) [inline tfjs-layers/src/engine/topology.ts:1034]: what if the array only contains one mask?

File: tfjs-layers/src/layers/nlp/multihead_attention.ts [hunk 2]
```diff
@@ -813,12 +813,20 @@ export class MultiHeadAttention extends Layer {
...
+      }
+      if (valueMask != null) {
+        const mask = valueMask.expandDims(1); // Shape is [B, 1, S]
+        autoMask = autoMask ? logicalAnd(autoMask, mask) : mask;
+      }
       if (useCausalMask) {
         // the shape of the causal mask is [1, T, S]
         const mask = this.computeCausalMask(query, value);
-        autoMask = mask;
+        autoMask = autoMask ? logicalAnd(autoMask, mask) : mask;
```
Comment (pyu10055) [inline tfjs-layers/src/layers/nlp/multihead_attention.ts:828]: how is this associated with the Topology computeMask logic?

File: tfjs-layers/src/engine/topology.ts (hunk 1 above)
Comment (mattsoulanille) [reply tfjs-layers/src/engine/topology.ts:1034]: I think that should be an error. If there's only one mask for all the tensors, it should be returned as a `tensor` instead of a `[tensor]`.

Comment (mattsoulanille) [reply tfjs-layers/src/engine/topology.ts]: Actually, keras does not seem to broadcast masks at all. Each output tensor needs its own mask:
https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L2893-L2898

File: tfjs-layers/src/layers/nlp/multihead_attention.ts (hunk 2 above)
Comment (mattsoulanille) [reply tfjs-layers/src/layers/nlp/multihead_attention.ts:828]: Earlier layers (Embedding) have computeMask called to compute the mask for their output tensors. This layer uses those masks.

File: tfjs-layers/src/engine/topology.ts [hunk 3]
```diff
@@ -1642,3 +1660,29 @@ export function getSourceInputs(
...
+                          ): tensors is Tensor | Tensor[] {
+  let noneAreSymbolic = true;
+  for (const tensor of generic_utils.toList(tensors)) {
+    if (tensor instanceof SymbolicTensor) {
+      noneAreSymbolic = false;
+      break;
+    }
+  }
+  return noneAreSymbolic;
+}
```
Comment (mattsoulanille) [inline tfjs-layers/src/engine/topology.ts:1688]: I moved this here so I could write them as type guards (`tensors is Tensor...`)

File: tfjs-layers/src/layers/nlp/multihead_attention_test.ts [hunk 4]
```diff
@@ -188,101 +188,6 @@ describe('MultiHeadAttention', () => {
     expectTensorsNotClose(queryKernel, outputKernel, 1e-6);
   });
 
-  describeMathCPU('High Dimensional Attention', () => {
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/multihead_attention_test.ts:191]: Moved

File: tfjs-layers/src/layers/nlp/multihead_attention_test.ts [hunk 5]
```diff
@@ -298,52 +203,60 @@ describe('MultiHeadAttention', () => {
...
+        Array<boolean[]>(5).fill([true, false, false]),
+        [[true, true, true]].concat(
+          Array<boolean[]>(4).fill([false, false, false]))
+      ]);
+    });
+
     /**
      * Test that the value and causal masks are taken into account.
      */
-    function testValueMask(testcaseName: string, useCausalMask: boolean) {
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/multihead_attention_test.ts:305]: Refactored to explicitly declare both tests instead of using a loop to declare them.

File: tfjs-layers/src/layers/nlp/multihead_attention_test.ts [hunk 6]
```diff
@@ -482,6 +395,101 @@ describe('MultiHeadAttention', () => {
   // TODO(pforderique): Test serialization.
 });
 
+describeMathCPU('High Dimensional Attention', () => {
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/multihead_attention_test.ts:398]: Moved some tests from above to here.

**Review by fengwuyao — APPROVED** [review]

**Review by pyu10055 — APPROVED** [review]
